toolchain/check/import_ref.cpp

```diff
@@ -987,4 +1019,44 @@ auto LoadImportRef(Context& context, SemIR::InstId inst_id, SemIRLoc loc)
   }
 }
 
+// Imports the impl `import_impl_id` from the imported IR `import_ir`.
+static auto ImportImpl(Context& context, SemIR::ImportIRId import_ir_id,
+                       const SemIR::File& import_ir,
+                       SemIR::ImplId import_impl_id) -> void {
+  // Resolve the imported impl to a local impl ID.
+  ImportRefResolver resolver(context, import_ir_id);
+  const auto& import_impl = import_ir.impls().Get(import_impl_id);
+  auto self_id = resolver.ResolveType(import_impl.self_id);
+  auto constraint_id = resolver.ResolveType(import_impl.constraint_id);
+
+  // Import the definition if the impl is defined.
+  // TODO: Do we need to check for multiple definitions?
+  auto impl_id = context.impls().LookupOrAdd(self_id, constraint_id);
+  auto& impl = context.impls().Get(impl_id);
+  if (import_impl.is_defined()) {
+    // TODO: Create a scope for the `impl` if necessary.
+    // TODO: Consider importing the definition_id.
+
+    impl.witness_id = context.AddImportRef(
+        {.ir_id = import_ir_id, .inst_id = import_impl.witness_id});
+  }
+}
+
+// TODO: This doesn't belong in this file. Consider moving the import resolver
+// and this file elsewhere.
+auto ImportImpls(Context& context) -> void {
```

Comment:
Should there be a TODO for handling indirect imports?

i.e.:

```
library "a" api;

interface HasF {
  fn F();
}
```

```
library "b" api;

import library "a";

class C {}
impl C as HasF
```

```
library "c" api;

import library "b";

alias D = C;
```

```
library "d" api;

import library "a";
import library "c";

fn G(c: D) {
  c.(HasF.F)();
}
```

Reply:
The `impl` handling here should already work: library "c" will contain the `impl C as HasF` too. But it doesn't work for unrelated reasons -- library "d" sees two different `HasF` interfaces, one from library "a" and one from library "c" (copied from "b", copied from "a") -- so I don't think I can write a test that checks this, unfortunately.